vortex-buffer/src/bit/buf_mut.rs

```diff
@@ -0,0 +1,108 @@
+use crate::{buffer_mut, BitBuffer, BufferMut};
+
+/// A mutable bitset buffer that allows random access to individual bits for set and get.
+///
+///
+/// # Example
+/// ```
+/// use vortex_buffer::BitBufferMut;
+///
+/// let mut bools = BitBufferMut::new_unset(10);
+/// bools.set_value(9, true);
+/// for i in 0..9 {
+///    assert!(!bools.value(i));
+/// }
+/// assert!(bools.value(9));
+///
+/// // Freeze into a new bools vector.
+/// let bools = bools.freeze();
+/// ```
+///
+/// See also: [`crate::BitBuffer`].
+pub struct BitBufferMut {
+    buffer: BufferMut<u64>,
+    len: usize,
+}
+
+impl BitBufferMut {
+    /// Create a new mutable buffer with requested `len` and all bits set to `true`.
+    pub fn new_set(len: usize) -> Self {
+        Self {
+            buffer: buffer_mut![u64::MAX; len.div_ceil(64)],
+            len,
+        }
+    }
+
+    /// Create a new mutable buffer with requested `len` and all bits set to `false`.
+    pub fn new_unset(len: usize) -> Self {
+        Self {
+            buffer: buffer_mut![0u64; len.div_ceil(64)],
+            len,
+        }
+    }
+
+    /// Create a new mutable buffer of the requested length, backed by uninitialized memory.
+    ///
+    /// This is a performance optimization for cases where we know only some of the bits in the
+    /// memory will be read, for example if you want to store a sparse bitset where all reads first
+    /// get masked via another bitset.
+    ///
+    /// # Safety
+    ///
+    /// The caller must initialize the sparse bitset, and be sure to guard against accidentally
+    /// reading uninitialized bits as if they were initialized.
+    pub unsafe fn new_uninit(len: usize) -> Self {
+        let words = len.div_ceil(64);
+        let mut buffer = BufferMut::with_capacity(words);
+        // SAFETY: enforced by the caller.
+        unsafe { buffer.set_len(words) };
+
+        Self { buffer, len }
+    }
+
+    /// Get the value at the requested index.
+    pub fn value(&self, index: usize) -> bool {
+        assert!(index < self.len, "index {index} exceeds len {}", self.len);
+
+        let word = self.buffer[index / 64];
+        let bit = word & (1 << (index % 64));
+
+        bit != 0
+    }
+
+    /// Set the bit at `index` to `value`.
+    pub fn set_value(&mut self, index: usize, value: bool) {
```

Comment:
Can we have `set_bit` and `unset_bit` so we can have a branchless API? This function can then call into those.

Reply:
We also want unsafe versions of getters and setters, see https://docs.rs/arrow-buffer/54.2.0/src/arrow_buffer/util/bit_util.rs.html#54

Reply:
yep